source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md

```diff
@@ -0,0 +1,194 @@
+
+Once we have such a hash (no worries, we've got one prepared), we can use Web3's `fromAscii()` utils to convert that hash to bytes and then send it off using our Smart Contract's `createPost()` method. We can then subscribe to the events we're emitting and check its return value like this:
+
+<pre class="javascript"><code>...
+const ipfsHash = 'Qmc5gCcjYypU7y28oCALwfSvxCBskLuPKWpK4qpterKC7z';
+
+contract('DReddit', () => {
+  ...
+  it ('should be able to create a post and receive it via contract event', async () => {
+    let receipt = await DReddit.methods.createPost(web3.utils.fromAscii(ipfsHash)).send();
+    const event = receipt.events.NewPost;
+    postId = event.returnValues.postId;
+    assert.equal(web3.utils.toAscii(event.returnValues.description), ipfsHash);
+  });
+});</code></pre>
+
+Notice that we're using `async/await` here because Embark's Smart Contract instance methods return promises. The same can be done without promises as well, it's just a syntactical difference at this point. Running `embark test` should result in two passing tests now!
+
+## Testing correctness of data
+
+Another good test case would be to check if the stored data such as the description bytes, the owner etc.  resolve back to the correct data. Notice that this is slightly different from what we're testing in our previous test - there we're testing the description bytes emitted by the `NewPost` event. To test this we take advantage of the `postId` created in the previous test, which is available globally now, to fetch the stored post. We then perform a similar check as in the previous test. We also want to test if the owner data of the post is correct, but for that we need to get access to the account that created the post in the first place.
+
+Luckily wallet accounts can be easily accessed as they are emitted by Embark's `config()` function. All we have to do is attaching a resolution handler to `config()` and storing the emitted value:
+
+<pre class="javascript"><code>...
+let accounts = [];
+
+config({
+  contracts: {
+    DReddit: {}
+  }
+}, (err, _accounts) => {
+  accounts = _accounts;
+});</code></pre>
+
+Having that in place, our next test could look something like this:
+
+<pre class="javascript"><code>contract('DReddit', () => {
+  ...
+  it ('post should have correct data', async () => {
+    const post = await DReddit.methods.posts(postId).call();
+    assert.equal(web3.utils.toAscii(post.description), ipfsHash);
+    assert.equal(post.owner, accounts[0]);
+  });
+});</code></pre>
+
+## Testing `canVote()`
+
+Alright, next up let's quickly test if our `canVote()` method works the way as expected. As voting on posts that don't exist should never work, we will simply call `canVote()` on a post id that doesn't exist. This test is pretty straight forward:
+
+<pre class="javascript"><code>contract('DReddit', () => {
```

Comment:
IMO, you don't have to put the `contract(...` everywhere. After one or to example, users probably got the gist of it.

Reply:
Yea, okay will roll back on that one.